glslc/src/main.cc

```diff
@@ -165,35 +177,54 @@ int main(int argc, char** argv) {
         }
       }
     } else if (arg == "-c") {
-      compiler.SetIndividualCompilationMode();
+      compiler.SetIndividualCompilationFlag();
     } else if (arg == "-E") {
-      compiler.SetPreprocessingOnlyMode();
+      compiler.options().SetPreprocessingOnlyMode();
+      compiler.SetPreprocessingOnlyFlag();
     } else if (arg == "-S") {
-      compiler.SetDisassemblyMode();
+      compiler.options().SetDisassemblyMode();
+      compiler.SetDisassemblyFlag();
     } else if (arg.starts_with("-D")) {
       const size_t length = arg.size();
       if (length <= 2) {
         std::cerr << "glslc: error: argument to '-D' is missing" << std::endl;
       } else {
         const string_piece argument = arg.substr(2);
-        size_t name_length = argument.find_first_of('=');
-        const string_piece name = argument.substr(0, name_length);
-        if (name.starts_with("GL_")) {
+        // Get the exact length of the macro string.
+        size_t equal_sign_loc = argument.find_first_of('=');
+        size_t name_length = equal_sign_loc != shaderc_util::string_piece::npos
+                                 ? equal_sign_loc
+                                 : (arg.size() - strlen("-D"));
+        const string_piece name_piece = argument.substr(0, name_length);
+        size_t argument_length = argument.size() - (argument.back() == '=' ? 1 : 0);
```

Comment:
This is confusing and complex.  I suggest removing it and checking for value absence in a different way -- see below.